tools/cli/lib/api.c

```diff
@@ -671,6 +671,211 @@ TDNFCliRepoSyncCommand(
+    } else if (strcmp(tag, "evr") == 0) {
+        pszVal = pPkgInfo->pszEVR;
+    } else if (strcmp(tag, "sourcename") == 0) {
+        pszVal = pPkgInfo->pszSourcePkg;
+    } else if (strcmp(tag, "size") == 0) {
+        pszVal = pPkgInfo->pszFormattedDownloadSize;
+    } else if (strcmp(tag, "downloadsize") == 0) {
+        pszVal = pPkgInfo->pszFormattedDownloadSize;
+    } else if (strcmp(tag, "installsize") == 0) {
+        pszVal = pPkgInfo->pszFormattedSize;
+    } else if (strcmp(tag, "sourcerpm") == 0) {
+        pszVal = pPkgInfo->pszSourcePkg;
+    } else if (strcmp(tag, "description") == 0) {
+        pszVal = pPkgInfo->pszDescription;
+    } else if (strcmp(tag, "summary") == 0) {
+        pszVal = pPkgInfo->pszSummary;
+    } else if (strcmp(tag, "license") == 0) {
+        pszVal = pPkgInfo->pszLicense;
+    } else if (strcmp(tag, "url") == 0) {
+        pszVal = pPkgInfo->pszURL;
+    } else if (strcmp(tag, "provides") == 0) {
+        dwError = parse_depResult(pPkgInfo->ppszDependencies[0], &pszVal);
+    } else if (strcmp(tag, "obsoletes") == 0) {
+        dwError = parse_depResult(pPkgInfo->ppszDependencies[1], &pszVal);
+    } else if (strcmp(tag, "conflicts") == 0) {
+        dwError = parse_depResult(pPkgInfo->ppszDependencies[2], &pszVal);
+    } else if (strcmp(tag, "requires") == 0) {
+        dwError = parse_depResult(pPkgInfo->ppszDependencies[3], &pszVal);
+    } else {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (pszVal) {
+        *pszRet = strdup(pszVal);
+    } else {
+        *pszRet = strdup("(null)");
+    }
+
+cleanup:
+    return dwError;
+error:
+    goto cleanup;
+}
+
+/* similar semantics as snprintf() */
+uint32_t format_string(char **out, size_t *len, const char *format,
+                  uint32_t (*tag_handler)(const char *, void *, char **), void *data)
+{
+    const char *p;
+    char *q;
+    int l = 0;
+    int nLength = 0;
+    uint32_t dwError = 0;
+    int buffer_bytes = 20;
+
+    p = format;
+    q = *out;
+    while (*p) {
+        if (*p == '%' && p[1] == '{') {
+            const char *t = &p[2];
+            char *tag, *value;
+            size_t tag_len;
+
+            while(*t && *t != '}') t++;
+
+            if (!*t)
+                return -1;
+            tag_len = t - &p[2];
+            tag = (char *)malloc(tag_len+1);
+            strncpy(tag, &p[2], tag_len);
+            tag[tag_len] = 0;
+            value = NULL;
+ 
+            dwError = tag_handler(tag, data, &value);
+            if (dwError == ERROR_TDNF_INVALID_PARAMETER) {
+                pr_err("Package doesn't have attribute %s\n", tag);
+            }
+            BAIL_ON_TDNF_ERROR(dwError);
+
+            if ((*len <= strlen(value))  || ((*len >= strlen(*out)) && (strlen(value) > (*len - strlen(*out))))) {
+                nLength = strlen(*out) + strlen(value) + buffer_bytes;
+                dwError = TDNFReAllocateMemory(nLength, (void **)out);
```

Comment:
Nit: this works, but usually for this pattern the space for reallocation is doubled. That would may safe a few repeated reallocations. If something needs n+1 bytes, it's most likely okay to allocate 2*n bytes.

Reply:
Done